File: osero_copy.py
# osero.py
class osero:
    def __init__(self, size = 8):
        self.size = size
        self.directions = [
            (-1, 0), (1, 0), (0, -1), (0, 1),
            (-1, -1), (-1, 1), (1, -1), (1, 1)
        ]
        board_init = [[0 for _ in range(size)] for _ in range(size)]
        mid = size // 2
        board_init[mid-1][mid-1] = 1
        board_init[mid][mid] = 1
        board_init[mid-1][mid] = -1
        board_init[mid][mid-1] = -1
        self.board = board_init
        self.history = []
        self.turn = 0

    def load_board_from_file(self, filename):
        """ファイルからボードを読み込む"""
        board = []
        with open(filename, 'r') as file:
            for line in file:
                row = [int(cell) for cell in line.strip().split()]  # 空白を分割して整数に変換
                board.append(row)
        return board

    def print_board_human(self): # 人間用
        """人が読み取りやすいマップを表示"""
        # "B" = -1 (黒), "W" = 1 (白), "." = 空き
        board = [['B' if cell == -1 else 'W' if cell == 1 else '.' for cell in row] for row in self.board]
        # 盤面の表示
        print("Othello Board(human):")
        print("Turn:", self.turn + 1)
        print("  c " + " ".join(map(str, list(range(self.size)))))
        print("r + " + "- " * len(board[0]) + "+")
        for i, row in enumerate(board):
            print(str(i) + " | " + " ".join(row) + " |")  # map(str, row) を削除
        print("  + " + "- " * len(board[0]) + "+")

    def print_board_human_can_place(self, piece): # 人間用
        """人が読み取りやすいマップを表示しながら置ける場所を示す"""
        # "B" = -1 (黒), "W" = 1 (白), "." = 空き
        board = [['B' if cell == -1 else 'W' if cell == 1 else '.' for cell in row] for row in self.board]
        # 盤面の表示
        print("Othello Board(human):")
        print("Turn:", self.turn + 1)
        print("  c " + " ".join(map(str, list(range(self.size)))))
        print("r + " + "- " * len(board[0]) + "+")
        for i, row in enumerate(board):
            row_str = " ".join(row)
            # 駒を置ける場所を示す
            for j in range(len(row)):
                if self.can_place_piece(i, j, piece):
                    row_str = row_str[:j*2] + '*' + row_str[j*2+1:]
            print(str(i) + " | " + row_str + " |")
        print("  + " + "- " * len(board[0]) + "+")

    def print_board_pc(self): # PC用
        """PC用の盤面を表示(内部データをそのまま表示)"""
        print("Othello Board(PC):")
        for row in self.board:
            print(row)

    def can_place_piece(self, row, col, piece):
        board = self.board
        rows = len(board)
        cols = len(board[0])
        opponent_piece = 1 if piece == -1 else -1
        if row < 0 or row >= rows or col < 0 or col >= cols or board[row][col] != 0:
            return False
        for dr, dc in self.directions:
            r, c = row + dr, col + dc
            found_opponent = False
            while 0 <= r < rows and 0 <= c < cols and board[r][c] == opponent_piece:
                found_opponent = True
                r += dr
                c += dc
            if found_opponent and 0 <= r < rows and 0 <= c < cols and board[r][c] == piece:
                return True
        return False

    def will_can_piece(self, piece):
        """置けるところをリストで返す"""
        li_can = []
        for r in range(self.size):
            for c in range(self.size):
                if self.can_place_piece(r, c, piece):
                    li_can.append((r, c))
        return li_can

    def line_change_piece(self, row, col, piece): # Github Copilotのから提供 Thanks! (自己改良済み)
        """置いた駒に応じて裏返す"""
        # 盤面のサイズ
        board = self.board
        rows = len(board)
        cols = len(board[0])

        # 各方向を確認して裏返す
        for dr, dc in self.directions:
            r, c = row + dr, col + dc
            path = []  # 挟んだ駒の座標を記録するリスト
            while 0 <= r < rows and 0 <= c < cols and board[r][c] != 0 and board[r][c] != piece:
                path.append((r, c))
                r += dr
                c += dc

            # 挟んだ駒がある場合、裏返す
            if 0 <= r < rows and 0 <= c < cols and board[r][c] == piece:
                for pr, pc in path:
                    self.board[pr][pc] = piece
                    # 履歴に記録
                    self.history[-1]["row"].append(pr)
                    self.history[-1]["col"].append(pc)

    def add_piece(self, row, col, piece):
        """盤面に駒を置く"""
        board = self.board
        if row < 0 or row >= len(board) or col < 0 or col >= len(board[0]):
            raise ValueError("Row or column out of bounds")
        if self.can_place_piece(row, col, piece):
            board[row][col] = piece
            # ここで履歴を切り詰める
            if len(self.history) > self.turn:
                self.history = self.history[:self.turn]
            self.history.append({"row": [row], "col": [col], "piece": piece})
            self.turn += 1
            self.line_change_piece(row, col, piece)
        else:
            print(row, col, piece)
            raise ValueError("Invalid move")

    def isnot_finished(self):
        """ゲームが終了しているか確認"""
        # 盤面に空きがあるか確認
        for row in self.board:
            if 0 in row:
                return True
        return False

    def count_pieces(self):
        """盤面にある駒の数をカウント"""
        # どちらが多いか
        count = {1: 0, -1: 0, 0: 0}
        for row in self.board:
            for cell in row:
                if cell in count:
                    count[cell] += 1
        return count

def main():
    """オセロ実行メイン関数 """
    import random
    import time
    size = int(input("Enter board size (default 8): ") or 8)
    piece = int(input("Enter your piece (1 or -1)(default 1): ") or 1)
    if piece not in [1, -1]:
        print("Invalid piece. Please enter 1 or -1")
        return "Invalid piece start"

    maps = []
    di = {1:0, -1:0}
    game_over = 0
    n = int(input("Enter number of games to play (default 100): ") or 100)
    for j in range(n):
        board = osero(size)
        turn_piece = [1, -1][j % 2]
        pass_count = 0
        while board.isnot_finished():
            can_li = board.will_can_piece(piece=turn_piece)
            if not can_li:
                pass_count += 1
                if pass_count >= 2:
                    game_over += 1
                    print(f"Game {j+1} over.")
                    break
                turn_piece = 1 if turn_piece == -1 else -1
                continue
            else:
                pass_count = 0
            try:
                row, col = random.choice(can_li)
                board.add_piece(row, col, turn_piece)
                turn_piece = 1 if turn_piece == -1 else -1
            except ValueError as e:
                print(e)
            except KeyboardInterrupt:
                print("\nGame interrupted by user.")
                return "Game interrupted"
            except Exception as e:
                print(f"An error occurred: {e}")
                print("Please contact the developer.")
        count = board.count_pieces()
        x = 1 if count[1] > count[-1] else -1
        try:
            maps.append((board, count))
            di[x] += 1
        except:
            pass
        if j % (n // 10) == 0:
            print(f"Progress: {j * 100 // n}%")
    print("Game results:")
    print(di)
    input("Press Enter to exit...")
    return 0

if __name__ == "__main__":
    n = main()
    print("Exit code:", n)

File: osero_cpp/osero_copy.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <string>
#include <fstream>
#include <sstream>
#include <random>
#include <ctime>
#include <map>
#include <algorithm>

// オセロのクラス定義
class Osero {
public:
    int size;
    std::vector<std::pair<int, int>> directions;
    std::vector<std::vector<int>> board;
    std::vector<std::map<std::string, std::vector<int>>> history;
    int turn;

    Osero(int size_ = 8) : size(size_), turn(0) {
        // 8方向
        directions = {
            {-1, 0},  // 上
            {1, 0},   // 下
            {0, -1},  // 左
            {0, 1},   // 右
            {-1, -1}, // 左上
            {-1, 1},  // 右上
            {1, -1},  // 左下
            {1, 1}    // 右下
        };

        board = std::vector<std::vector<int>>(size, std::vector<int>(size, 0));
        int mid = size / 2;
        // 初期配置
        board[mid - 1][mid - 1] = 1;
        board[mid][mid] = 1;
        board[mid - 1][mid] = -1;
        board[mid][mid - 1] = -1;
    }

    std::vector<std::vector<int>> load_board_from_file(const std::string& filename) {
        std::vector<std::vector<int>> loaded_board;
        std::ifstream infile(filename);
        std::string line;
        while (std::getline(infile, line)) {
            std::vector<int> row;
            std::istringstream iss(line);
            int cell;
            while (iss >> cell) {
                row.push_back(cell);
            }
            loaded_board.push_back(row);
        }
        return loaded_board;
    }

    void print_board_human() const {
        std::cout << "Othello Board(human):" << std::endl;
        std::cout << "Turn: " << turn + 1 << std::endl;
        std::cout << "  c ";
        for (int i = 0; i < size; ++i) std::cout << i << " ";
        std::cout << std::endl;
        std::cout << "r + ";
        for (int i = 0; i < size; ++i) std::cout << "- ";
        std::cout << "+" << std::endl;
        for (int i = 0; i < size; ++i) {
            std::cout << i << " | ";
            for (int j = 0; j < size; ++j) {
                char ch = '.';
                if (board[i][j] == -1) ch = 'B';
                else if (board[i][j] == 1) ch = 'W';
                std::cout << ch << " ";
            }
            std::cout << "|" << std::endl;
        }
        std::cout << "  + ";
        for (int i = 0; i < size; ++i) std::cout << "- ";
        std::cout << "+" << std::endl;
    }

    void print_board_human_can_place(int piece) const {
        std::cout << "Othello Board(human):" << std::endl;
        std::cout << "Turn: " << turn + 1 << std::endl;
        std::cout << "  c ";
        for (int i = 0; i < size; ++i) std::cout << i << " ";
        std::cout << std::endl;
        std::cout << "r + ";
        for (int i = 0; i < size; ++i) std::cout << "- ";
        std::cout << "+" << std::endl;
        for (int i = 0; i < size; ++i) {
            std::cout << i << " | ";
            for (int j = 0; j < size; ++j) {
                char ch = '.';
                if (board[i][j] == -1) ch = 'B';
                else if (board[i][j] == 1) ch = 'W';
                if (can_place_piece(i, j, piece)) ch = '*';
                std::cout << ch << " ";
            }
            std::cout << "|" << std::endl;
        }
        std::cout << "  + ";
        for (int i = 0; i < size; ++i) std::cout << "- ";
        std::cout << "+" << std::endl;
    }

    void print_board_pc() const {
        std::cout << "Othello Board(PC):" << std::endl;
        for (const auto& row : board) {
            for (auto cell : row) std::cout << cell << " ";
            std::cout << std::endl;
        }
    }

    bool can_place_piece(int row, int col, int piece) const {
        if (row < 0 || row >= size || col < 0 || col >= size || board[row][col] != 0)
            return false;
        int opponent_piece = (piece == -1) ? 1 : -1;

        for (const auto& dir : directions) {
            int r = row + dir.first, c = col + dir.second;
            bool found_opponent = false;
            while (r >= 0 && r < size && c >= 0 && c < size && board[r][c] == opponent_piece) {
                found_opponent = true;
                r += dir.first;
                c += dir.second;
            }
            if (found_opponent && r >= 0 && r < size && c >= 0 && c < size && board[r][c] == piece)
                return true;
        }
        return false;
    }

    void line_change_piece(int row, int col, int piece) {
        int rows = size, cols = size;
        for (const auto& dir : directions) {
            int r = row + dir.first, c = col + dir.second;
            std::vector<std::pair<int, int>> path;
            while (r >= 0 && r < rows && c >= 0 && c < cols && board[r][c] != 0 && board[r][c] != piece) {
                path.push_back({r, c});
                r += dir.first;
                c += dir.second;
            }
            if (r >= 0 && r < rows && c >= 0 && c < cols && board[r][c] == piece) {
                for (const auto& p : path) {
                    board[p.first][p.second] = piece;
                    history.back()["row"].push_back(p.first);
                    history.back()["col"].push_back(p.second);
                }
            }
        }
    }

    void add_piece(int row, int col, int piece) {
        if (row < 0 || row >= size || col < 0 || col >= size)
            throw std::invalid_argument("Row or column out of bounds");
        if (can_place_piece(row, col, piece)) {
            board[row][col] = piece;
            if ((int)history.size() > turn)
                history.resize(turn);
            history.push_back({{"row", {row}}, {"col", {col}}, {"piece", {piece}}});
            turn += 1;
            line_change_piece(row, col, piece);
        } else {
            throw std::invalid_argument("Invalid move");
        }
    }

    bool isnot_finished() const {
        for (const auto& row : board) {
            if (std::find(row.begin(), row.end(), 0) != row.end())
                return true;
        }
        return false;
    }

    std::map<int, int> count_pieces() const {
        std::map<int, int> count{{1, 0}, {-1, 0}, {0, 0}};
        for (const auto& row : board) {
            for (auto cell : row) {
                if (count.find(cell) != count.end())
                    count[cell]++;
            }
        }
        return count;
    }

    int piece_back() {
        if (turn <= 0)
            throw std::invalid_argument("No moves to undo");
        std::cout << "piece_back" << std::endl;
        const auto& last_move = history[turn - 1];
        int row = last_move.at("row")[0], col = last_move.at("col")[0];
        int piece = last_move.at("piece")[0];
        board[row][col] = 0;
        auto row_it = last_move.at("row").begin() + 1;
        auto col_it = last_move.at("col").begin() + 1;
        for (; row_it != last_move.at("row").end() && col_it != last_move.at("col").end(); ++row_it, ++col_it) {
            board[*row_it][*col_it] = -piece;
        }
        turn -= 1;
        return -piece;
    }

    int piece_forward() {
        if (turn >= (int)history.size())
            throw std::invalid_argument("No moves to redo");
        std::cout << "piece_forward" << std::endl;
        const auto& last_move = history[turn];
        int row = last_move.at("row")[0], col = last_move.at("col")[0];
        int piece = last_move.at("piece")[0];
        board[row][col] = piece;
        auto row_it = last_move.at("row").begin() + 1;
        auto col_it = last_move.at("col").begin() + 1;
        for (; row_it != last_move.at("row").end() && col_it != last_move.at("col").end(); ++row_it, ++col_it) {
            board[*row_it][*col_it] = piece;
        }
        turn += 1;
        return -piece;
    }

    std::vector<std::pair<int, int>> piece_proposal(int piece) const {
        std::vector<std::pair<int, int>> proposals;
        for (int r = 0; r < size; ++r) {
            for (int c = 0; c < size; ++c) {
                if (can_place_piece(r, c, piece)) {
                    proposals.push_back({r, c});
                }
            }
        }
        return proposals;
    }
};

int main() {
    std::srand((unsigned int)std::time(nullptr));
    int size;
    std::cout << "Enter board size (default 8): ";
    std::string size_in;
    std::getline(std::cin, size_in);
    size = size_in.empty() ? 8 : std::stoi(size_in);

    int piece;
    std::cout << "Enter your piece (1 or -1): ";
    std::string piece_in;
    std::getline(std::cin, piece_in);
    piece = piece_in.empty() ? 1 : std::stoi(piece_in);
    if (piece != 1 && piece != -1) {
        std::cout << "Invalid piece. Please enter 1 or -1." << std::endl;
        std::cout << "Invalid piece start" << std::endl;
        return 1;
    }

    std::vector<std::tuple<Osero, std::map<int, int>>> maps;
    std::map<int, int> di{{1, 0}, {-1, 0}};
    int n;
    std::cout << "Enter number of games to play (default 10): ";
    std::string n_in;
    std::getline(std::cin, n_in);
    n = n_in.empty() ? 10 : std::stoi(n_in);

    for (int j = 0; j < n; ++j) {
        Osero board(size);
        int turn_piece = (j % 2 == 0) ? 1 : -1;
        int pass_count = 0;
        while (board.isnot_finished()) {
            auto count = board.count_pieces();
            auto can_li = board.piece_proposal(turn_piece);
            if (can_li.empty()) {
                turn_piece = (turn_piece == -1) ? 1 : -1;
                board.turn += 1;
                pass_count += 1;
                if (pass_count >= 2) {
                    std::cout << "Both players cannot move. Game over." << std::endl;
                    break;
                }
                continue;
            }
            pass_count = 0;
            try {
                std::pair<int, int> move = can_li[std::rand() % can_li.size()];
                board.add_piece(move.first, move.second, turn_piece);
                turn_piece = (turn_piece == -1) ? 1 : -1;
            } catch (const std::exception& e) {
                std::cout << e.what() << std::endl;
            }
        }
        auto count = board.count_pieces();
        int x = (count[1] > count[-1]) ? 1 : -1;
        try {
            maps.push_back(std::make_tuple(board, count));
            di[x]++;
        } catch (...) {
        }
        if (j % std::max(1, n / 10) == 0) {
            std::cout << "Progress: " << j * 100 / n << "%" << std::endl;
        }
    }
    std::cout << "Game results:" << std::endl;
    // 結果表示省略
    std::cout << "1: " << di[1] << ", -1: " << di[-1] << std::endl;
    std::cout << "Press Enter to exit..." << std::endl;
    std::cin.get();
    return 0;
}